### PR TITLE
Install vscode-ripgrep at runtime

### PR DESCRIPTION
**Pull request by thieman**
This was fun :smile: @roblourens if you're still interested I'd love any thoughts you have. Also let me know if this seems to fix things for you, I think it's working for me but I've only got Ubuntu to test on

Potential fix for #1

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,7 +25,7 @@
     "sublime"
   ],
   "engines": {
-    "vscode": "^1.30.0"
+    "vscode": "^1.31.0"
```
Comment (thieman) [inline package.json:28]: Necessary to pick up `globalStoragePath`

File: package.json [hunk 2]
```diff
@@ -113,6 +113,15 @@
...
+    "@types/request": "^2.48.1",
+    "@types/tar-fs": "^1.16.1",
+    "github-releases": "^0.4.1",
+    "gunzip-maybe": "^1.4.1",
+    "progress": "^2.0.0",
+    "request": "^2.88.0",
+    "rimraf": "^2.6.1",
+    "tar-fs": "^2.0.0",
+    "vscode-ripgrep": "^1.2.5",
+    "yauzl": "^2.8.0"
```
Comment (thieman) [inline package.json:125]: Installed all of `vscode-ripgrep`'s dependencies so we can require pieces of it and have everything work

File: src/ripgrep.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    case "win32":
+      opts.version = "0.10.0-patch.0";
+      opts.arch = os.arch();
+      break;
+    case "linux":
+      opts.arch = os.arch();
+      break;
+    default:
+      throw new Error("Unknown platform: " + opts.platform);
+  }
```
Comment (thieman) [inline src/ripgrep.ts:80]: I cribbed most of this function from vscode-ripgrep directly, please let me know if that's kosher and if not I can rewrite it

File: src/extension.ts [hunk 4]
```diff
@@ -1,8 +1,16 @@
 import * as vscode from "vscode";
 import * as commands from "./commands";
 import { BetterSearchProvider } from "./providers";
+import { ensureRipgrepInstalled } from "./ripgrep";
+
+export let context: vscode.ExtensionContext;
```
Comment (thieman) [inline src/extension.ts:6]: I didn't see a reason not to treat this like a global, am I missing something?

**Comment by roblourens** [comment]
Glad you got it to work! `download.js` in vscode-ripgrep isn't really intended to be API, so I can't guarantee that it won't break in another vscode-ripgrep version. You might try to see if you can just invoke `postinstall.js` which is what runs after the module is installed with npm, that could be safer if you ever want to update vscode-ripgrep. If you wanted to you could also download vscode-ripgrep into the extension's node_modules folder, so that you could basically require() it normally instead of building the executable path yourself. But this approach should also work just fine. I'll test it on Mac once it's published!

**Comment by thieman** [comment]
Understand totally about the compatibility concern, the vscode-ripgrep version is hardcoded so at least it won't break before we intentionally update it

>  You might try to see if you can just invoke postinstall.js

I looked at this but my impression was that I couldn't use it directly without submitting some changes to it. One thing that jumped out was `process.exit` in the error handler which I'm guessing would have crashed the entire extension

Doesn't sound like I can make it worse than it already is so I'll get this merged and publish a new version, thanks again for your help 🎉

**Comment by roblourens** [comment]
Oh yeah, please don't call process.exit in the extension host 👍

**Comment by thieman** [comment]
0.0.2 is published :tada:

**Comment by roblourens** [comment]
Hm, it's not working for me now, I can try to debug it later

**Comment by thieman** [comment]
Agh it's an escaping problem, the path on Mac has a space in it 😡

Welp here comes another version

**Comment by thieman** [comment]
Okay 0.0.3 is out and I'm feeling good about this one :smile:

**Comment by thieman** [comment]
Two snags I'll just mention:

1. VSCode let me install the extension even when I was running an older version. If you're not on at least 1.31 it's not going to work, though
2. The new "update extensions without reloading" thing isn't working very well for me, I updated the extension and it wouldn't work until I reloaded the window

**Pull request merged**